kernel/posix/pthread.c

```diff
@@ -0,0 +1,531 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <atomic.h>
+#include "ksched.h"
+#include "wait_q.h"
+
+#define PTHREAD_INIT_FLAGS	PTHREAD_CANCEL_ENABLE
+#define PTHREAD_CANCELED	((void *) -1)
+
+static const pthread_attr_t init_pthread_attrs = {
+	.priority = K_LOWEST_APPLICATION_THREAD_PRIO,
+	.stack = NULL,
+	.stacksize = 0,
+	.flags = PTHREAD_INIT_FLAGS,
+	.delayedstart = K_NO_WAIT,
+	.schedpolicy = SCHED_RR,
+	.detachstate = PTHREAD_CREATE_JOINABLE,
+	.initialized = true,
+};
+
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_thread_pool, sizeof(struct posix_thread),
```

Comment:
Nothing is returned to this memory pool; we'll still run out of pthread handles.

I remember threads having a callback function that got called whenever they were destroyed.  Can't we used that to return something to the pool?  It might be only for threads that are aborted, which is fine; pthread_exit() can schedule a call to k_mem_pool_free() (it can't call it directly).

Reply:
Currently I am not returning back to mem pool becuase we need to keep pthread object to handle 
pthread_join requests. If we will free the space then pthread_join will have issue.

But I agree we need to figure out a way to return back to mem pool on pthread termination. I can explore it further. But even when k_thread is destroyed, we may need to access pthread object such as retval, pthtread state, statue_mutex to address pthread_join.